Mathlib/Data/Option/Basic.lean

```diff
@@ -3,28 +3,366 @@ Copyright (c) 2017 Mario Carneiro. All rights reserved.
+
+-/
 
 namespace Option
 
-theorem some_injective (α : Type _) : Function.Injective (@some α) :=
-  fun _ _ => some_inj.mp
+variable {α β γ δ : Type _}
+
+theorem coe_def : (fun a => ↑a : α → Option α) = some :=
+  rfl
+
+#align option.get_or_else Option.getD
+
+@[simp]
+theorem getD_coe (x y : α) : Option.getD (↑x) y = x :=
+  rfl
+
+theorem coe_get {o : Option α} (h : o.isSome) : ((Option.get _ h : α) : Option α) = o :=
+  Option.some_get h
+
+theorem eq_of_mem_of_mem {a : α} {o1 o2 : Option α} (h1 : a ∈ o1) (h2 : a ∈ o2) : o1 = o2 :=
+  h1.trans h2.symm
+
+theorem Mem.left_unique : Relator.LeftUnique ((· ∈ ·) : α → Option α → Prop) :=
+fun _ _ _=> mem_unique
+
+theorem some_injective (α : Type _) : Function.Injective (@some α) := fun _ _ => some_inj.mp
 
 /-- `option.map f` is injective if `f` is injective. -/
 theorem map_injective {f : α → β} (Hf : Function.Injective f) : Function.Injective (Option.map f)
-| none, none, _ => rfl
-| some a₁, some a₂, H => by rw [Hf (Option.some.inj H)]
+  | none, none, _ => rfl
+  | some a₁, some a₂, H => by rw [Hf (Option.some.inj H)]
+
+@[simp]
+theorem map_comp_some (f : α → β) : Option.map f ∘ some = some ∘ f :=
+  rfl
+
+@[simp]
+theorem none_bind' (f : α → Option β) : none.bind f = none :=
+  rfl
+
+@[simp]
+theorem some_bind' (a : α) (f : α → Option β) : (some a).bind f = f a :=
+  rfl
+
+theorem bind_eq_some' {x : Option α} {f : α → Option β} {b : β} :
+    x.bind f = some b ↔ ∃ a, x = some a ∧ f a = some b :=
+  by cases x <;> simp
+
+theorem bind_eq_none' {o : Option α} {f : α → Option β} :
+    o.bind f = none ↔ ∀ b a, a ∈ o → b ∉ f a := by
+  simp only [eq_none_iff_forall_not_mem, mem_def, bind_eq_some, not_exists, not_and, iff_self]
+
+-- FIXME: there is no global `mjoin`
+-- theorem join_eq_join : mjoin = @join α :=
+--   funext fun x => by rw [mjoin, bind_id_eq_join]
+
+theorem bind_eq_bind {α β : Type _} {f : α → Option β} {x : Option α} : x >>= f = x.bind f :=
+  rfl
+
+theorem map_coe {α β} {a : α} {f : α → β} : f <$> (a : Option α) = ↑(f a) :=
+  rfl
+
+@[simp]
+theorem map_coe' {a : α} {f : α → β} : Option.map f (a : Option α) = ↑(f a) :=
+  rfl
+
+/-- `option.map` as a function between functions is injective. -/
+theorem map_injective' : Function.Injective (@Option.map α β) := fun f g h =>
+  funext fun x => some_injective _ <| by simp only [← map_some', h]
+
+@[simp]
+theorem map_inj {f g : α → β} : Option.map f = Option.map g ↔ f = g :=
+  map_injective'.eq_iff
+
+attribute [simp] map_id
+
+@[simp]
+theorem map_eq_id {f : α → α} : Option.map f = id ↔ f = id :=
+  map_injective'.eq_iff' map_id
+
+theorem map_comm {f₁ : α → β} {f₂ : α → γ} {g₁ : β → δ} {g₂ : γ → δ} (h : g₁ ∘ f₁ = g₂ ∘ f₂)
+  (a : α) :
+    (Option.map f₁ a).map g₁ = (Option.map f₂ a).map g₂ := by rw [map_map, h, ← map_map]
+
+section Pmap
+
+variable {p : α → Prop} (f : ∀ a : α, p a → β) (x : Option α)
+
+-- Can't simp tag this anymore because `pbind` simplifies
+-- @[simp]
```

Comment:
I suggest all comments like this get a common phrase, probably "Porting note:" at the beginning.

However I think we're just removing all the simp lemmas where the Lean 4 simp already simplifies the left hand side because it looks inside definitions by pattern matching itself.

Reply:
Does your comment mean:
- Add `Porting note:`, leave `-- @[simp]`
- Add `Porting note:`, remove `-- @[simp]`
- Have neither

I've done option 1 so far.

Reply:
Sorry, it was meant to be two separate suggestions:
1. Definitely add "Porting note" to any comment that is about porting, and will stay in the PR post-merge.
2. For this class of simp lemmas, I *think* (Mario may correct!) that it is okay to just drop these @[simp] annotations without leaving a comment.